Bundle KLU solver in PyBaMM wheel

tlestang opened the pull request:
This PR primarily addresses the compilation and distribution of the idaklu solver.
The idaklu solver is based on a python C++ extension that must be compiled before the solver can be used, i.e. before import idaklu can be run.

At the moment, the idaklu solver is not part of the PyBaMM package (it appears nowhere in the setup.py). Should a user want to use the idaklu solver, they must clone the GitHub repository and compile the solver themselves.
Furthermore, the resulting compile module (e.g. `idaklu.cpython-XYm-x86_64-linux-gnu.so`) is not properly installed alongside the package and instead just dumped in the compilation directory.

## Summary of changes
- The `idaklu` module is now declared as a python extension in the `setup.py`. This enables the shipping of the `idaklu` module as as part of the PyPI package. In practice, there are now _no extra-steps_  for a user wishing to use the idaklu solver:
```
$ pip install pybamm
$ python -c "import pybamm; pybamm.have_idaklu()"
True
```
This is valid for both Linux, macosx and Windows within the WSL.
- On Windows (without WSL), PyBaMM can be pip installed, but does not come with the `idaklu` solver.
- The installation of `scikits.odes` has been moved to a separate script `pybamm/install_odes.py` and declared as an entry point in the `setup.py`. This means that, following the installation of PyBaMM, users can invoke
```
$ pybamm_install_odes [--install-sundials]
```
To install `scikits.odes`.
- For developpers wishing to work on/use the `idaklu` solver, a convenience script is provided that lays the groundwork for the compilation (i.e. downloads and installs SuiteSparse and sundials). Additional compilation/installtion notes are provided.
- In the current state of `develop`, package data like `CITATIONS.txt`, `version` of the content of `pybamm/input/parameters/` are not consistently added to the pybamm distribution. The `setup.py` has been modified and a `MANIFEST.in` added to address this problem.
- Compilation (since it contains C++) of the PyBaMM wheel distribution can be made compliant with the manylinux policy ([PEP 513](https://www.python.org/dev/peps/pep-0513/)), see issue #898. To produced the wheels this PR
adds a script `build-wheel.sh` that sould be used within one of the manylinux docker images. This script could be adapted to a Continuous Delivery setup, see #855.

## For reviewers
This PR addresses both the user side and the developper side.
To start off with, you could check that everything works as it should.

_Until this PR is merged, the updated PyBaMM is hosted on TestPyPI instead of PyPI_.
_In the following,_ `pip install pybamm` _should be replaced by_
```
pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple pybamm
```
_The above can sometimes fail in recent versions of Ubuntu (https://github.com/pypa/pip/issues/7486)._
_In this case_
```
# Get PyBaMM from TestPyPI (no dependencies)
pip install --no-deps --index-url https://test.pypi.org/simple/ pybamm 
# Install deps from regular PyPI
pip -install -r requirements.txt
```
_Note that at the moment no PyBaMM wheel is available for MacOS, simply because I don't have access to a mac. If any mac user would be willing to spend an hour with me to build a wheel and upload it to TestPyPI, that would be super useful._

- **As a user**, you should be able to:
  - [ ] `pip install pybamm` completely independently from the GitHub
repository.
  - [ ] import and use the `idaklu` solver. 
  - [ ] enable the `scikits.odes` solver by using the command `$ pybamm_install_odes --install-sundials`.
  - [ ] Find a solution to your problem by reading the install notes.
- **As a developper**, from a freshly cloned repo, you should be able to:
  - [ ] Install pybamm from source using `pip install .`, `pip install -e.`, `python setup.py install`..
  - [ ] Check the installation logs `PyBaMM/setup.log` (or `your/virtual/env/site-packages/pybamm/setup.log` if you didn't use the `-e` switch or `develop`).
  - [ ] Compile the `idaklu` module following the instructions in `compiling-KLU.md`

The central file in this PR is `setup.py`. The main change compared to the current `develop` is 
```
setup(
...
	ext_modules=ext_modules,
        cmdclass={
        "build_ext": CMakeBuild.CMakeBuild,
        "bdist_wheel": bdist_wheel,
        "install": CustomInstall,
    },
```
This declares the `idaklu.cpp` file as an _extension module_ and replace the command executed by `python setup.py build_ext` by a CMake build defined in `CMakeBuild.py`.
The `CMakeLists.txt` has been modified so that CMake should find SuiteSparse and sundials in common
locations. It now uses a `FindSUNDIALS.cmake` module.
The `build_ext` command can take two options to specify the location of both libraries if they are installed in an unusual location. See installation notes `compiling-KLU.md`.
Most of the time, the `build_ext` is called from the commands `python setup.py install` or `python setup.py bdist_wheel` (to create a wheel). In order for the sustom `build_ext` to receive the two options I defined two custom `install` and `bdist_wheel` from which the custom `build_ext` receives the option's values from.
See [the distutils docs on adding new commands](https://docs.python.org/3.7/distutils/extending.html) and this [post](https://jichu4n.com/posts/how-to-add-custom-build-steps-and-commands-to-setuppy/).

Independently from the `setup.py`, the script that downloads and builds SuiteSparse and sundials is
`scripts/setup_KLU_module_build.py`.
It could be used by a developper wisihing to compile the `idaklu` solver, but is also use in `build-wheel.sh` to build PyBaMM wheels.
For more details about wheels and stuff like `auditwheel`, see #898.



## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)


# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

tinosulzer commented:
Great stuff @tlestang , thanks! I will let someone else who is more familiar with this stuff review

tinosulzer commented:
btw, I am working on a fuel cell model where the only solver that works is the scikit dae one, so it would be good if we could keep it

tlestang commented:
Oh ok, apologies. I've just reverted the `travis.yaml` to include the scikits.odes case and added a script to install scikits.odes.
After a `pip install pybamm`, a user should be able to install scikits.odes using
```
pybamm_install_odes --install-sundials
```
This will download and install sundials, and install scikits.odes. The `--install-sundials` tells the script to download an install the sundials.
I'll update the installation docs next

tlestang commented:
Thanks @martinjrobins 

### Developer install
#### Missing `wheel` module
wheel is installed by default is any venv created by `virtualenv`, but not the `venv` module. I think `virtualenv` should be recommended as its maintained by PyPA and superseeds `venv`, see https://virtualenv.pypa.io/en/latest/
#### Missing `libklu.so.1`
Did this happen despite installing KLU and sundials following the instructions in `compiling-KLU.md`? Did you use the convenience script to download and install them? 

### User install
pip can get pybamm's dependencies and that's a Ubuntu bug, see [https://github.com/pypa/pip/issues/7486](pypa#7486).
Install pybamm from TestPyPI without dependencies using the `--no-deps` switch
```
pip install --no-deps --index-url https://test.pypi.org/simple/ pybamm 
```
And install its deps will a regular `pip install` from PyPI. e.g.
```
pip -r requirements.txt
```

tlestang commented:
@martinjrobins I fixed the install on Ubuntu 18.04. Turns out the Ubuntu repos provide CMake 3.10 which is not compatible with the way the FindSUNDIALS.cmake module looks for the Sundials.
I've updated the cmake module to make it work with CMake < 3.12

tinosulzer commented on MANIFEST.in:
```diff
@@ -0,0 +1,86 @@
+include pybamm/input/parameters/lithium-ion/separators/separator_Chen2020/README.md
+include pybamm/input/parameters/lithium-ion/separators/separator_Marquis2019/README.md
+include pybamm/version
+include pybamm/CITATIONS.txt
```
do new parameter sets automatically get added to this, or does it need to be done manually?

tlestang replied:
It's manual at the moment, and I have a script that goes through the directories and writes the `MANIFEST.in`.
This file is actually only required for including the parameters in the source distribution (the archive uploaded ti PyPI) and not in the built distribution (the wheel).
The data included in the wheel is defined in the setup.py (variable `pybamm_data`) and everything inside `input/` is included.

This is not good practice though, as python package distributions are meant to only ship the source code... As the number of parameters grow we should probably think of other ways to distribute parameters to users. Another repo? I'm planning to write an issue about this.

tinosulzer commented on compiling-KLU.md:
```diff
@@ -0,0 +1,157 @@
+# PyBaMM developer install - The KLU sparse solver
+If you wish so simulate large systems such as the 2+1D models, we recommend employing a
+sparse solver.
```
I think casadi solver is "sparse" so this could be confusing

tlestang replied:
What is the main motivation for using the KLU solver then? Is it significantly faster?
Could the above sentence be replaced by
```
If you wish so simulate large systems such as the 2+1D models, we recommend employing the KLU sparse solver provided by the Sundials library.
```

tinosulzer replied:
It's not faster either, just sometimes works when casadi doesn't ... so I guess
```
If you wish to try a different DAE solver, you can install the KLU sparse solver provided by the Sundials library.
```

tlestang commented:
For reasons I don't fully understand yet, importing the `idaklu` solver fails on Ubuntu 18.04, reporting not found `libsundials_sunmatrixsparse.so.3`.
 For the record, both SuiteSparse and Sundials are installed in `KLU_module_deps` in their respective directory. One can check that these directory are listed in the `DT_RUNPATH` entry of the python module
```
$ readelf -d idaklu.cpython-36m-x86_64-linux-gnu.so 

Dynamic section at offset 0x62c50 contains 29 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libsundials_ida.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libsundials_sunlinsolklu.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000f (RUNPATH)              Library runpath: [/tmp/PyBaMM/KLU_module_deps/sundials5/lib:/tmp/PyBaMM/KLU_module_deps/SuiteSparse-5.6.0/lib]
```
However only `ida` and `sunlinsolklu` are listed here.
Upgrading the dynamic linker `ld.so` from `2.30` (default on Ubuntu 18.04) to `2.31` (default on Debian 10) seems to solve the problem.

As an alternative, the path to both SuiteSparse and Sundials location must now be added to the `LD_LIBRARY_PATH`. In addition, both libraries are now installed in `~/.local/` instead of the PyBaMM project directory.
